pkg/services/pluginsintegration/plugincontext/plugincontext.go

```diff
@@ -11,68 +11,83 @@ import (
-	pluginSettingsService pluginsettings.Service) *Provider {
+	dataSourceService datasources.DataSourceService, pluginSettingsService pluginsettings.Service,
+	keyProvider KeyProvider) *Provider {
 	return &Provider{
+		keyProvider:           keyProvider,
 		cacheService:          cacheService,
 		pluginStore:           pluginStore,
-		dataSourceCache:       dataSourceCache,
 		dataSourceService:     dataSourceService,
 		pluginSettingsService: pluginSettingsService,
-		logger:                log.New("plugincontext"),
 	}
 }
 
 type Provider struct {
 	cacheService          *localcache.CacheService
 	pluginStore           plugins.Store
-	dataSourceCache       datasources.CacheService
 	dataSourceService     datasources.DataSourceService
 	pluginSettingsService pluginsettings.Service
-	logger                log.Logger
+	keyProvider           KeyProvider
 }
 
 // Get allows getting plugin context by its ID. If datasourceUID is not empty string
 // then PluginContext.DataSourceInstanceSettings will be resolved and appended to
 // returned context.
-func (p *Provider) Get(ctx context.Context, pluginID string, user *user.SignedInUser) (backend.PluginContext, bool, error) {
-	return p.pluginContext(ctx, pluginID, user)
+func (p *Provider) Get(ctx context.Context, pluginID string, user *user.SignedInUser) (backend.PluginContext, error) {
+	plugin, exists := p.pluginStore.Plugin(ctx, pluginID)
+	if !exists {
+		return backend.PluginContext{}, ErrPluginNotFound
+	}
+
+	pCtx := backend.PluginContext{
+		OrgID:    user.OrgID,
+		PluginID: pluginID,
+		User:     adapters.BackendUserFromSignedInUser(user),
+	}
+
+	if plugin.IsApp() {
+		appSettings, err := p.appInstanceSettings(ctx, pluginID, user)
+		if err != nil {
+			return backend.PluginContext{}, err
+		}
+		pCtx.AppInstanceSettings = appSettings
+		//pCtx.Key = p.keyProvider.AppKey(ctx, pluginID, user.OrgID)
+	}
+
+	return pCtx, nil
 }
 
 // GetWithDataSource allows getting plugin context by its ID and PluginContext.DataSourceInstanceSettings will be
 // resolved and appended to the returned context.
-func (p *Provider) GetWithDataSource(ctx context.Context, pluginID string, user *user.SignedInUser, ds *datasources.DataSource) (backend.PluginContext, bool, error) {
-	pCtx, exists, err := p.pluginContext(ctx, pluginID, user)
+func (p *Provider) GetWithDataSource(ctx context.Context, pluginID string, user *user.SignedInUser, ds *datasources.DataSource) (backend.PluginContext, error) {
+	pCtx, err := p.Get(ctx, pluginID, user)
 	if err != nil {
-		return pCtx, exists, err
+		return backend.PluginContext{}, err
 	}
 
 	datasourceSettings, err := adapters.ModelToInstanceSettings(ds, p.decryptSecureJsonDataFn(ctx))
 	if err != nil {
-		return pCtx, exists, fmt.Errorf("%v: %w", "Failed to convert datasource", err)
+		return pCtx, err
 	}
 	pCtx.DataSourceInstanceSettings = datasourceSettings
+	//pCtx.Key = p.keyProvider.DataSourceKey(ctx, datasourceSettings)
```

Comment:
same here?

Reply:
Yeah sorry I should mention that changes are required to the SDK to introduce a new field (for example `Key` to the plugin context) - https://github.com/grafana/grafana-plugin-sdk-go/blob/main/backend/common.go#L128. This is purely to demonstrate what it might look like

Reply:
okay, so what this PR is doing, if I understand that correctly is:

 - Simplify the usage of pluginContext
 - Define a KeyService
 - (To be enabled) Enable the KeyService within the pluginContext

That makes sense :+1: Maybe the only comment I would do is to use `CacheKeyService` for the new service, so it's clear what is its purpose.